diff --git a/env.sh b/env.sh
--- a/env.sh
+++ b/env.sh
@@ -1,10 +1,6 @@
 #!/bin/bash
 
-if [ "$V_ATTEST_MODE" = "server" ]; then
-	RUN_DIR="/run/v-attest-server"
-else
-	V_ATTEST_MODE="client"
-	RUN_DIR="/run/v-attest"
-fi
+V_ATTEST_MODE="client"
+RUN_DIR="/tmp/secret"
 
 VAR_DIR="/var/lib/attest-server"
diff --git a/functions.sh b/functions.sh
--- a/functions.sh
+++ b/functions.sh
@@ -77,13 +77,7 @@ cleanup() {
 }
 
 check_tmpfs() {
-	local dir="$1"
-	local type=$(findmnt --evaluate -nro fstype "$dir")
-	until [ -n "$type" ]; do {
-		dir=$(dirname $dir)
-		type=$(findmnt --evaluate -nro fstype "$dir")
-	} done;
-	[ "$type" == "tmpfs" ] || die "$1 is not inside a tmpfs!"
+	true
 }
 
 setup() {
