diff --git a/Makefile.include b/Makefile.include
--- a/Makefile.include
+++ b/Makefile.include
@@ -45,11 +45,15 @@ $(call debug_shell, $(CC) -Wall -Werror $2 $1 $3 -o /dev/null && echo yes || ech
 endef
 
 define find_dependency
-$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) $(PKG_CONFIG) --exists $1 && echo yes || echo no)
+$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) \
+	$(if $(PKG_CONFIG_SYSROOT_DIR),PKG_CONFIG_SYSROOT_DIR=$(PKG_CONFIG_SYSROOT_DIR),) \
+	$(PKG_CONFIG) --exists $1 && echo yes || echo no)
 endef
 
 define dependency_version
-$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) $(PKG_CONFIG) --modversion $1 2>/dev/null)
+$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) \
+	$(if $(PKG_CONFIG_SYSROOT_DIR),PKG_CONFIG_SYSROOT_DIR=$(PKG_CONFIG_SYSROOT_DIR),) \
+	$(PKG_CONFIG) --modversion $1 2>/dev/null)
 endef
 
 define has_dependency # $1: dependency, $2: action/target
@@ -57,9 +61,13 @@ $(if $(findstring $(strip $1),yes), $(strip $2))
 endef
 
 define dependency_cflags
-$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) $(PKG_CONFIG) --cflags $1 2>/dev/null)
+$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) \
+	$(if $(PKG_CONFIG_SYSROOT_DIR),PKG_CONFIG_SYSROOT_DIR=$(PKG_CONFIG_SYSROOT_DIR),) \
+	$(PKG_CONFIG) --cflags $1 2>/dev/null)
 endef
 
 define dependency_ldflags
-$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) $(PKG_CONFIG) --libs --static $1 2>/dev/null)
+$(call debug_shell, $(if $(PKG_CONFIG_LIBDIR),PKG_CONFIG_LIBDIR=$(PKG_CONFIG_LIBDIR),) \
+	$(if $(PKG_CONFIG_SYSROOT_DIR),PKG_CONFIG_SYSROOT_DIR=$(PKG_CONFIG_SYSROOT_DIR),) \
+	$(PKG_CONFIG) --libs --static $1 2>/dev/null)
 endef
